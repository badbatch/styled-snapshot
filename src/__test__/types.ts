import { Func } from "@repodog/types";
import { ComponentType } from "enzyme";
import { ReactElement, ReactFragment, ReactPortal, Ref } from "react";
import { FlattenSimpleInterpolation } from "styled-components";
import { ContextConsumerElement, ContextProviderElement, ForwardRefElement, MemoElement } from "../types";

export interface ComponentProps {
  callback?: Func;
  Component?: ComponentType<{}>;
  consumer?: ContextConsumerElement;
  css?: FlattenSimpleInterpolation;
  decorator?: ForwardRefElement;
  element?: ReactElement;
  forwardRef?: ForwardRefElement;
  fragment?: ReactFragment;
  innerRef?: Ref<HTMLDivElement>;
  memo?: MemoElement;
  portal?: ReactPortal;
  provider?: ContextProviderElement;
  renderProp?: (...args: any[]) => ReactElement;
  styled?: ComponentType<{}>;
  styledFunctionComponent?: ReactElement;
  styledStyledComponent?: ReactElement;
  styles?: FlattenSimpleInterpolation;
  value?: string;
}
