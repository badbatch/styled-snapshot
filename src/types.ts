import { Func, ObjectMap } from "@repodog/types";
import { ShallowWrapper } from "enzyme";
import {
  ComponentClass,
  ComponentType,
  Consumer,
  ConsumerProps,
  ExoticComponent,
  ForwardRefExoticComponent,
  FunctionComponent,
  MemoExoticComponent,
  PropsWithChildren,
  PropsWithoutRef,
  Provider,
  ProviderProps,
  ReactElement,
  ReactNode,
  ReactPortal,
  RefAttributes,
  RefForwardingComponent,
} from "react";
import { AnyStyledComponent } from "styled-components";

export type ComponentTypeElement =
  | ReactElement<PropsWithChildren<{}>, ComponentClass>
  | ReactElement<PropsWithChildren<{}>, FunctionComponent>;

export type ContextConsumerElement = ReactElement<ConsumerProps<any>, Consumer<any>>; // tslint:disable-line no-any

// tslint:disable-next-line no-any
export type ContextProviderElement = ReactElement<ProviderProps<any>, Provider<any> & { _context: PrivateContext }>;

export type DomElement = ReactElement<PropsWithChildren<{}>, string>;

export type ForwardRefComponent = ForwardRefExoticComponent<
  PropsWithoutRef<Element> & RefAttributes<PropsWithChildren<{}>>
> & {
  readonly $$typeof: symbol;
  render: RefForwardingComponent<Element, PropsWithChildren<{}>>;
};

export type ForwardRefElement = ReactElement<PropsWithChildren<{}>, ForwardRefComponent>;

export type FragmentElement = ReactElement<PropsWithChildren<{}>, ExoticComponent<{ children?: ReactNode }>>;

export type MandatoryUnwrapElement =
  | ContextConsumerElement
  | ContextProviderElement
  | ForwardRefElement
  | FragmentElement
  | MemoElement
  | MemoExoticComponent<ComponentType>
  | ReactPortal;

export type MemoElement = ReactElement<PropsWithChildren<{}>, MemoExoticComponent<ComponentType>>;

export interface PrivateContext {
  Consumer: Consumer<any>; // tslint:disable-line no-any
  Provider: Provider<any>; // tslint:disable-line no-any
}

export type ValidElement = ComponentTypeElement | MandatoryUnwrapElement | DomElement;

export type ReactTreeVisitor = (node: ObjectMap) => void;

export interface StyledSnapshotConfig {
  elementsToUnwrap?: string[];
  reactTreeVisitor?: ReactTreeVisitor;
  unwrapCustomizer?: UnwrapCustomizer;
}

export type UnwrapCustomizer = (element: ValidElement) => ValidElement | ReactNode;

export type ComponentTree = ShallowWrapper<
  {
    children?: ReactNode;
  },
  Readonly<{}>,
  React.Component<{}, {}, any> // tslint:disable-line no-any
>;

export interface SerializedTree {
  $$typeof: symbol;
  children: SerializedTree[] | ReactElement[];
  node: ReactElement;
  props: ObjectMap;
  type: string;
}

export type TreeNode = SerializedTree;

export type ExtractedContexts = Map<React.ExoticComponent<React.ConsumerProps<any>>, any>; // tslint:disable-line no-any

export interface UnwrapElementResult {
  contexts: ExtractedContexts;
  element: ComponentTypeElement;
}

export interface SCSerializedTree {
  $$typeof: symbol;
  children: null;
  node: AnyStyledComponent;
  props: SCSerializedTreeProps;
  type: string;
}

export interface SCSerializedTreeProps {
  forwardedComponent: SCForwardRefComponent;
}

export type SCForwardRefComponent = ForwardRefComponent & { componentStyle: SCComponentStyle; displayName: string };

export interface SCComponentStyle {
  rules: Array<string | number | Func>;
}
