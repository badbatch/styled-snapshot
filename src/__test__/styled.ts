import styled from "styled-components";

export const StyledDiv = styled.div`
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  box-sizing: border-box;
  position: relative;
`;
