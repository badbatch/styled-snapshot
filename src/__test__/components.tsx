import { ObjectMap } from "@repodog/types";
import PropTypes from "prop-types";
import React, { Component, ComponentType, forwardRef, memo } from "react";
import { createPortal } from "react-dom";
import styled, { ThemeContext, ThemeProvider, withTheme } from "styled-components";
import { StyledDiv, StyledItem, StyledList } from "./styled";
import { ComponentProps } from "./types";

export class ClassComponent extends Component<ComponentProps> {
  public static displayName = "ClassComponent";

  public static propTypes = {
    Component: PropTypes.elementType,
  };

  // tslint:disable-next-line prefer-function-over-method
  public render() {
    return this.props.children;
  }
}

export const ContextComponent = () => (
  <ThemeContext.Provider value="dark">
    <ContextConsumerComponent />
  </ThemeContext.Provider>
);

ContextComponent.displayName = "ContextComponent";

export const ContextConsumerComponent = () => (
  <ThemeContext.Consumer>
    {value => (
      <ClassComponent value={value}>
        <FunctionComponent />
      </ClassComponent>
    )}
  </ThemeContext.Consumer>
);

ContextConsumerComponent.displayName = "ContextConsumerComponent";

export const ContextProviderComponent = () => (
  <ThemeContext.Provider value="dark">
    <ClassComponent>
      <FunctionComponent />
    </ClassComponent>
  </ThemeContext.Provider>
);

ContextProviderComponent.displayName = "ContextProviderComponent";

export const ForwardRefComponent = forwardRef<HTMLDivElement, ComponentProps>((props, ref) => (
  <ClassComponent {...props} innerRef={ref}>
    <FunctionComponent />
  </ClassComponent>
));

ForwardRefComponent.displayName = "ForwardRefComponent";

export const FunctionComponent = (props: ComponentProps) => <StyledDiv {...props} />;
FunctionComponent.displayName = "FunctionComponent";

FunctionComponent.propTypes = {
  Component: PropTypes.elementType,
};

export const MemoComponent = memo(FunctionComponent);
MemoComponent.displayName = "MemoComponent";

export const portal = createPortal(
  <ClassComponent>
    <FunctionComponent />
  </ClassComponent>,
  document.createElement("div"),
);

export const PortalComponent = () => portal;
PortalComponent.displayName = "PortalComponent";

export const SCComponent = () => (
  <StyledDiv>
    <ClassComponent>
      <FunctionComponent />
    </ClassComponent>
    <FunctionComponent />
    <StyledList>
      <StyledItem>One</StyledItem>
      <StyledItem>Two</StyledItem>
      <StyledItem>Three</StyledItem>
      <StyledItem>Four</StyledItem>
    </StyledList>
  </StyledDiv>
);

SCComponent.displayName = "SCComponent";

export const SCContextComponent = () => (
  <ThemeProvider theme={{ type: "light" }}>
    <WithThemeClassComponent />
  </ThemeProvider>
);

SCContextComponent.displayName = "SCContextComponent";

export const WithThemeClassComponent = withTheme<ComponentType<any>>(ClassComponent); // tslint:disable-line no-any

export const SerializeComponent = (props: ObjectMap = {}) => {
  return (
    <ClassComponent
      Component={FunctionComponent}
      callback={() => null}
      consumer={<ThemeContext.Consumer>{() => <div />}</ThemeContext.Consumer>}
      decorator={<WithThemeClassComponent />}
      element={<FunctionComponent {...props} />}
      forwardRef={<ForwardRefComponent />}
      fragment={<></>}
      memo={<MemoComponent />}
      portal={portal}
      provider={<ThemeContext.Provider value="" />}
      renderProp={() => <FunctionComponent {...props} renderProp={subProps => <div {...subProps} />} />}
      styled={StyledDiv}
      styledFunctionComponent={<StyledFunctionComponent />}
      styledStyledComponent={<StyledStyledComponent />}
    >
      <ForwardRefComponent />
      <FunctionComponent {...props} />
      <MemoComponent />
      <StyledFunctionComponent />
      <StyledList>
        {[1, 2, 3, 4].map(id => (
          <StyledStyledComponent key={id} />
        ))}
      </StyledList>
    </ClassComponent>
  );
};

SerializeComponent.displayName = "SerializeComponent";

export const StyledFunctionComponent = styled(FunctionComponent)``;

export const StyledStyledComponent = styled(StyledDiv)``;
