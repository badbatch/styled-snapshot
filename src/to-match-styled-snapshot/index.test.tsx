import { ObjectMap } from "@repodog/types";
import { has } from "lodash";
import React from "react";
import { ThemeProvider } from "styled-components";
import toMatchStyledSnapshot from ".";
import { SCComponent } from "../__test__/components";

describe("toMatchStyledSnapshot", () => {
  describe("when a React element is passed in", () => {
    const visitor = (node: ObjectMap) => {
      if (has(node, ["props", "theme"])) {
        node.props.theme = null;
      }
    };

    toMatchStyledSnapshot(
      <ThemeProvider theme={{ type: "light" }}>
        <SCComponent />
      </ThemeProvider>,
      {
        elementsToUnwrap: ["ThemeProvider"],
        reactTreeVisitor: visitor,
      },
    );
  });
});
