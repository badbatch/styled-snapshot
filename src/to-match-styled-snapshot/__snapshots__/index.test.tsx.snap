// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toMatchStyledSnapshot >> SCComponent >> component >> styled.div 1`] = `
"
{
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  box-sizing: border-box;
  position: relative;
}
"
`;

exports[`toMatchStyledSnapshot >> SCComponent >> component >> styled.li 1`] = `
"
{
  margin-left: 12px;

  position: absolute;
  padding: null;
  &:last-child {
    margin-left: 0;
  }
}
"
`;

exports[`toMatchStyledSnapshot >> SCComponent >> component >> styled.ul 1`] = `
"
{
  margin: 0;
  padding: 0;
}
"
`;

exports[`toMatchStyledSnapshot >> SCComponent >> component 1`] = `
<styled.div>
  <ClassComponent>
    <FunctionComponent />
  </ClassComponent>
  <FunctionComponent />
  <styled.ul>
    <styled.li>
      One
    </styled.li>
    <styled.li>
      Two
    </styled.li>
    <styled.li>
      Three
    </styled.li>
    <styled.li>
      Four
    </styled.li>
  </styled.ul>
</styled.div>
`;

exports[`toMatchStyledSnapshot >> SerializeComponent >> component >> Styled(styled.div) 1`] = `
"
{
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  box-sizing: border-box;
  position: relative;
}
"
`;

exports[`toMatchStyledSnapshot >> SerializeComponent >> component >> styled.ul 1`] = `
"
{
  margin: 0;
  padding: 0;
}
"
`;

exports[`toMatchStyledSnapshot >> SerializeComponent >> component 1`] = `
<ClassComponent
  Component={Symbol(FunctionComponent)}
  callback={[Function]}
  consumer={
    <Context.Consumer>
      [Function]
    </Context.Consumer>
  }
  decorator={<WithTheme(ClassComponent) />}
  element={<FunctionComponent />}
  forwardRef={<ForwardRefComponent />}
  fragment={<React.Fragment />}
  memo={<Memo(MemoComponent) />}
  portal={
    <Portal(ClassComponent)>
      <FunctionComponent />
    </Portal(ClassComponent)>
  }
  provider={
    <Context.Provider
      value=""
    />
  }
  renderProp={[Function]}
  styled={Symbol(styled.div)}
  styledFunctionComponent={<Styled(FunctionComponent) />}
  styledStyledComponent={<Styled(styled.div) />}
>
  <ForwardRefComponent />
  <FunctionComponent />
  <Memo(MemoComponent) />
  <Styled(FunctionComponent) />
  <styled.ul>
    <Styled(styled.div) />
    <Styled(styled.div) />
    <Styled(styled.div) />
    <Styled(styled.div) />
  </styled.ul>
</ClassComponent>
`;
